### PR TITLE
Highlight substrings correctly and initial exclusion support with `-`

### LINKED ISSUE #128923: Allow exact and excluded matches in fuzzy search.
In fuzzy searches, allow you to require an exact match by using double quotes and exclude terms using a minus sign.  Here are some examples:
- `"abc"` would match `abc`, but not `abac`
- `abc -def` would match `abc`, but not `abcdef`.

### PR DESCRIPTION
**Pull request by TylerLeonhardt**
I’ve been playing around with a few ways to hone in on the files you’re looking for in the cmd+P file search. Taking inspiration from Google & GitHub here’s what quoting and exclusion might look like:

* wrap a query piece in quotes and that substring must exist "foo"
* add a dash in front of a query piece and that substring must NOT exist -foo

The results:
| | |
|---|---|
| quotes-and-minuses | minus-vscode-repo |

(exclude query pieces can only be used if there’s another piece)
(“pieces” of a query are separated by spaces so asdf bsdf has 2 pieces)

Initially I felt like the exclusion query pieces weren't needed but then I used it more and found it really nice tacking on additional exclusions.

2 commits initially:
* initial implementation that highlights consecutive maches correctly - this makes it so when searching for `"26"` the highlight is 2021-9-**26** and not **2**021-9-2**6** (if that's hard to read, the 2nd one has highlights on the first 2 instead of the 2 next to the 6)
* initial support of `-foo` exclusion query parts

I would recommend looking at the first commit first and then the 2nd.

I haven't fixed up the tests yet but I was hoping to get some feedback from @bpasero before I do that to make sure I'm on the right track generally speaking.

This PR fixes #128923

### CODE REVIEW

File: src/vs/base/common/fuzzyScorer.ts [hunk 1]
```diff
@@ -460,12 +487,13 @@ function doScoreItemFuzzyMultiple(label: string, description: string | undefined
 function doScoreItemFuzzySingle(label: string, description: string | undefined, path: string | undefined, query: IPreparedQueryPiece, preferLabelMatches: boolean, fuzzy: boolean): IItemScore {
 
 	// Prefer label matches if told so or we have no description
-	if (preferLabelMatches || !description) {
+	if ((preferLabelMatches && !query.excludeQuery) || !description) {
```
Comment (TylerLeonhardt) [inline src/vs/base/common/fuzzyScorer.ts:490]: this is a bit tricky... but the idea is that exclude queries have to happen on the entire path.

File: src/vs/base/common/fuzzyScorer.ts [hunk 2]
```diff
@@ -831,6 +866,14 @@ function queryExpectsExactMatch(query: string) {
 	return query.startsWith('"') && query.endsWith('"');
 }
 
+/*
+ * If a query is wrapped in quotes, the user does not want to
+ * use fuzzy search for this query.
+ */
```
Comment (TylerLeonhardt) [inline src/vs/base/common/fuzzyScorer.ts:872]:
```suggestion
/*
 * If a query starts with a dash and isn't just a dash, the user does not want this
 * query to appear in results.
 */
```

**Comment by bpasero** [comment]
I am wondering how the highlighting is not working properly when using quotes for forcing contiguos matches, isn't that the same that is used for editor history matches, so I would expect we had a bug there already?

I suggest to split the PR up into fixing the highlighting and implementing the support for `-` so we can discuss both independently.

**Comment by bpasero** [comment]
Oh yeah, I can confirm the highlighting issue, good catch. Reproduces in stable too with entries from editor history:

image

**Review by bpasero — CHANGES_REQUESTED** [review]
High level feedback on the matching fix:
* rename `fuzzy: boolean` to `allowNonContiguousMatches` to make the intent clear (the current wording is confusing)
* not a big fan of bypassing the scorer matrix for when an exact match is searched, can we not use it but assign a score of 0 within the matrix when we are expecting exact matches and the current match is not contiguous? that way we are not breaking any ranking behaviour we have today already

**Comment by TylerLeonhardt** [comment]
Ok I've moved this to https://github.com/microsoft/vscode/pull/131292

**Comment by bpasero** [comment]
I am not sure whether we want to push forward on support for `-` but wouldnt the easiest solution for that just be this:
* find a leading `-` and mark the query with a boolean property (similar to how we deal with quotes)
* run the search normally as before, e.g. each part that is separated by space is a query on its own
* for any query that is this special one with a `-` the score is computed as normal but treated specially:
  * a `0` is ignored
  * a score greater `0` results in a score of 0

As such I would think the code changes are relatively small.

**Comment by TylerLeonhardt** [comment]
I agree, @bpasero. I think it'd be easy to add. Before I go and add `-` support, I wanna chat with @JacksonKearl and a few others to maybe do some thinking on alignment with our search experiences.

**Pull request closed without merging**